### ISSUE TITLE
Extensible disambiguation

### ISSUE THREAD

**Issue by krame505** [melt-umn/silver#193]
This got a bit long, so apologies for that, but I wanted to make sure that the background for this problem is clear. 

We currently do not have a way of resolving lexical ambiguities in a 'smart' way that allows extensions to add new terminals that share an ambiguity.  An extension may try to write a disambiguation function, for example the following that was considered for the template extension:
```
disambiguate TemplateIdentifier_t, TemplateTypeName_t, Identifier_t, TypeName_t
{
  pluck
    case lookupBy(stringEq, lexeme, head(context)) of
    | just(lh:templateIdentType_c()) -> TemplateIdentifier_t
    | just(lh:templateTypenameType_c()) -> TemplateTypeName_t
    | just(lh:identType_c()) -> Identifier_t
    | just(lh:typenameType_c()) -> TypeName_t
    | nothing() -> Identifier_t
    | it -> error(...)
    end;
}
```
The issues with this approach are twofold: First, we need to re-specify the host disambiguation logic (the 3rd and 4th lines of the case), which is ugly and a chance for errors to be introduced.  Secondly, and worse, if more terminals are added to the ambiguous set by another extension, then the whole function no longer applies and all terminals must be distinguished via transparent prefixes in an ugly way.  

A couple of things to note here:
* It is still possible for there to be multiple ambiguities for a terminal that must be resolved with different logic.
* All the cases in the above example (except for the error case) are mutually exclusive, so the order of cases doesn't matter.  

I am proposing that we add something along the lines of the following.
* A method to declare a 'disambiguation class' of lexically ambiguous terminals that are resolved according to a certain method:
```disambiguation class Identifier;```
* A method of specifying how a terminal is a member of a disambiguation class:
```
terminal Identifier_t /[A-Za-z_\$][A-Za-z_0-9\$]*/ lexer classes {Cidentifier}
disambiguate Identifier {
  case lookupBy(stringEq, lexeme, head(context)) of
    | just(lh:identType_c()) -> true
    | _ -> false
  end;
}
```
Here the body of the block is a Boolean expression that is true if that terminal is to be selected.  Note that all of these in a disambiguation class should be disjoint so that the process can be deterministic, but since arbitrary logic can be used that unfortunately can't be checked statically, so the exact preference between these would be undefined.  If no item is valid, then some sort of runtime error should be raised.  I'm also not 100% sure about this syntax, or if this should really be an action block, but that can be decided.  

There are a couple of ways that this could be implemented:
* By Silver, at composition time.  This would mean that declaring a parser would generate disambiguation functions for every combination of terminals in each disambiguation class.  This would blow up exponentially at 2^n - n, but would probably be an OK interim solution since we won't be having huge number of terminals in these classes.  Generating these disambiguation functions could happen similarly to how we generate those used for transparent prefixes.
* By Copper, statically and eventually at runtime.  I'm not real sure about how this would work, but it presumably wouldn't be much more complex than disambiguation functions.  Here we would also have the advantage of knowing what terminals actually conflict before generating everything.  

My suggestion, if this is implemented, is to implement it in Silver first before looking to add native support in Copper.  @tedinski @ericvanwyk any thoughts?

**Comment by krame505** [comment]
Also, I guess I wasn't very clear on the proposed semantics of disambiguation classes.  
A disambiguation class corresponds to a 'method' of resolving lexical conflicts for any conflicting subset of the class.  A terminal can be a member of more than one, if it conflicts with multiple different terminals in different situations calling for different methods of resolution, but no pair of terminals in one disambiguation class should both be members of another disambiguation class, as this could lead to an ambiguity of which method of disambiguation to use.

**Comment by tedinski** [comment]
My thoughts: I think I like the general idea. In particular, it's possible this could be a solution to a more general problem.

We presently frame marking terminals as somethings that only conflict with each other. However, there's another problem of marking terminals conflicting with existing identifiers. Presently, we treat them as keywords, which means if you have `table` as a marking terminal and a program that has a variable named `table`, you get conflicts this way. This (should? actually, idk) be fixable with prefixes just the same, but it's a bigger problem because it's a lot easier to have marking-identifier conflicts than marking-marking conflicts.

If we adopted a convention of introducing marking terminals as non-keywords, with a disambiguation class looking up the name, seeing if it's shadowed or not, and resolving on that basis, it would be a neat little solution to this problem.

HOWEVER, this is a direction that would complicate the composition process. I would recommend not pursuing this until after we have runtime composition of parse tables, and have fully integrated that into Silver (so we're composing extensions at runtime). IMO, that's a much more important and more urgent goal.

**Comment by krame505** [comment]
I was randomly thinking about this idea again and wanted to get my thoughts down - this concept can be made simpler, potentially avoiding issues with runtime composition.  We would introduce 2 independant new features to Silver/Copper:

* Terminal "guards": a boolean expression, yielding either true or false, to be evaluated whenever the regex for the terminal matches.  If the expression returns false then it is treated as if the regex didn't match - sort of like guards in pattern matching:
```
terminal Identifier_t /[A-Za-z_\$][A-Za-z_0-9\$]*/
  when
    case lookupBy(stringEq, lexeme, head(context)) of
      | just(lh:identType_c()) -> true
      | nothing() -> true
      | _ -> false
    end,
  lexer classes {Cidentifier};
```
This would have no effect on the compile-time analysis to find lexical ambiguities.  
* Ambiguous lexer classes - lexer classes for which Copper would allow unresolved lexical ambiguities between any subset of member terminals at compile time, trusting that invariants are in place to prevent more than exactly one member from matching (and causing a lexical ambiguity) at runtime:
```
lexer class Cidentifier ambiguous, submits to Ckeyword;
```

Both of these features could potentially be useful by themselves, however they can be used together to enable extensible runtime disambiguation.  If an extension, e.g. templates, defines a new identifier terminal in the `Cidentifier` lexer class, it will also define a new `IdentType` production, update the parsing `context` env when appropriate, and include a guard on the new terminal to ensure that it will only match if the identifier exists.  Other, similar extensions would be guaranteed not to conflict with this terminal, because their corresponding `IdentType` productions are not present in the `context` under that name.  

AFAIK this shouldn't be exceedingly difficult to implement now.  It also shouldn't cause an excessive burden on the runtime composition process, although I'm not totally clear on if we even would be checking lexical ambiguities when composing at runtime.  If we are, all that would be required is to verify whether perceived ambiguities are all contained in an ambiguous lexer class, and if so, ignore this conflict.  

Thoughts?

**Comment by krame505** [comment]
@ericvanwyk This is the issue that I mentioned yesterday in our discussion.  We would probably want to do something mostly in line with the second idea I outlined here, but we would need to think carefully about how this interacts with transparent prefixes - perhaps if we see a prefix followed by a terminal with a matching regex, then ignore the guard for that terminal?  Or something.

**Comment by ericvanwyk** [comment]
I think we can be more ambitious than this.  It feels like a bandaid when something more invasive might be warranted.

I think we can find a way to avoid "trusting" extension to satisfy the invariants you're thinking about.

But more importantly, I think ``inExtCode`` leaves some problems unsolved.  What about #include-ing a ``.xh`` file that is written using some extension (required by the extension defining the ``.xh`` file)?  Maybe we can direct Copper to scan/parse using on a subset of the LR parse table at times?  This would seem reasonable and could be used in interesting ways, I suspect.  

First you should read August's thesis to be sure you really understand what is going on in Copper.
   https://conservancy.umn.edu/handle/11299/95605
If you want to go deeper, then we could also read Joel Denny's thesis as it had some interesting ideas on precedence (for reserving keywords for example).
   https://tigerprints.clemson.edu/all_dissertations/519/

Also, should this discussion happen in the Copper repo?

**Comment by krame505** [comment]
Some further thoughts and summary of what we discussed earlier:

There are 2 fundamentally seperate issues in play here.  We want some way of enabling/disabling extension marking terminals, effectively including or excluding them from the context set of shiftable tokens.  We also want a way of generalizing disambiguation for classes of terminals.  

For both cases, we need a way of dealing with first-class "lexer class members" in Silver.  Currently inside disambiguation functions, we have special literals corresponding to terminals of a particular lexical ambiguity, however they can only be used within the function.  This can be made more general - make every lexer class a type in Silver, and various terminal identifiers would be values of this type, allowing terminal names to be written as an expression literal anywhere.  This would be a bit nicer if Silver had type classes, but for now we can deal with it as every lexer class introducing a new type.  Lexer class member literals would just be given type `a`, and we can error check that whatever the final type inferred here is actually some lexer class type.  

As a translation, these lexer class members would be java `String`s containing Silver's generated Copper name of the terminal.  Copper would also need to generate a function converting a terminal's Copper name `String` to its `int` value.  Initially we can do this with reflection on the generated `Terminals` enum, but this would be easy to change when we eventually add runtime composition.  

For the first issue, we want to add new statements to action blocks: `include foo;` or `exclude foo;` where `foo` is either the name of a lexer class or an expression of type `[<some lexer class>]`.  This would allow us to effectively set the list of marking terminals allowed at some point with a something like
```haskell
lexer class Marking;

-- In practice this would be a stack controlled by cpp tags
parser attribute enabledMarkingTerminals::[Marking];

concrete production includeExtension
top::Decl ::= 'include_extension' extName::Identifier_t ';'
{ ... }
action {
  enabledMarkingTerminals =
    enabledMarkingTerminals ++
    lookupExtTerminals(extName.lexeme); -- Some function with a collection attribute
  exclude Marking;
  include enabledMarkingTerminals;
}
```
This can be implemented by adding a 'mask' `BitSet` in Copper that is anded with the `pushable` `BitSet` that Copper passes to the scanner.  We would also need to have Copper generate the `BitSet`s corresponding to each lexer class - this would also require special handling eventually with runtime composition.  

For the second issue, we would want to do something similar to disambiguation classes as I originally proposed them, although having first-class lexer class member values means we wouldn't need to work in terms of boolean flags and associated invariants.  We could simply annotate a lexer class with an expression evaluating to a member of that class to make it a 'disambiguation class'.  For any lexical ambiguity (not resolved by an explicit disambiguation function) that is a subset of a disambiguation class, we would select the terminal indicate by the class if it is an option, or else raise an error as if none had matched.  For example:
```haskell
parser attribute context::[[Pair<String Cidentifer>]];

lexer class Cidentifier submits to Ckeyword,
  disambiguate {
    case lookupBy(stringEq, lexeme, head(context)) of
    | just(i) -> i
    | _ -> Identifier_t
    end;
  };
```

**Comment by tedinski** [comment]
I don't understand the point of `include`/`exclude`.

It looks like this can be summarized as:

1. We need a type for terminals, so within Silver we can build a data structure acting as a parse-time environment mapping strings to terminals for use inside a disambiguation action block.
2. We need Copper support for *something* like a "disambiguation class," allowing us an open set of terminals with ambiguities, where we can (typically) use an environment to either select a marking terminal, or a plain identifier.

**Comment by tedinski** [comment]
Okay, offline we had a chat, here are some thoughts:

I get where `include`/`exclude` are coming from now, but I suspect this may be a non-issue. The concern is namespace collisions for extensions between header files and with ordinary identifiers.

1. This might not be an extensible languages problem, it's a C problem. You already can't have `struct foo` in two headers, so C programmers already have to deal with this. I'm not sure the added problem being faced here is actually our problem.
2. The original idea behind transparent prefixes (I thought?) was that there was a "fully qualified name" for the marking terminal. In C++, for instance, you already aren't supposed to write header files that `using namespace std` you're supposed to fully-qualify everything in the header. We could have the same rule.
3. The rest of the problem, I think, reduces to the already existing "disambiguation class" problem.

If all extensions introduced keyword marking terminals via the disambiguation class, then we could just have each header file start with zero extensions in the parsing env, and declare which ones should be enabled locally with `#extension foo` or whatever, adding just those keyword marking terminals to the disambiguation environment.

I think that'd take care of it?

**Comment by krame505** [comment]
It looks like, to make this work for what we want in ableC, one more Silver-side piece is needed.  In order to have mutually recursive template structs and functions, some sort of forward declaration with no semantic meaning is needed to place the appropriate terminal identifier in the context parser attribute.  However, in the *actual* definition following the forward declaration still needs to be disambiguated as an identifier, for example:
```c++
template<> int a(void);
template<> int b(void) { a<>(); }
template<> int a(void) {}
```
the terminal `a` should still be disambiguated as `Identifier_t` in the last declaration, even though it is already in the context as `TerminalIdentifier_t`.  However, since `TerminalIdentifier_t` is not in the shiftable set at this point (only `Identifier_t`, `TypeName_t` and `TemplateTypeName_t` are), a syntax error results.  
The fix I propose for this is to also make the shiftable set available inside disambiguation classes, as a built-in parser attribute (similar to `lexeme`) of type `[TerminalId]`.  Since computing this list entails some overhead and may not be required by all disambiguation classes, it can be done lazily.  
If the terminal identifier looked up from the context is not currently shiftable, then we can instead return `Identifier_t` by default.  I'm going to take a stab at implementing this to see if it handles all the problem cases I have encountered.  This is a seperate issue from the one I found with MDA tests (https://github.com/melt-umn/copper/issues/37)
